### PR TITLE
promotions widget and page

### PR DESCRIPTION
**Pull request by eugenebiz**

### CODE REVIEW

File: Bizyhood/Core.php [hunk 1]
```diff
@@ -1033,10 +1070,129 @@ public function businesses_information($atts)
...
+        return;
+      }
+      
+      $a = shortcode_atts( array(
+        'bid'         => null,
+        'identifier'  => null
+      ), $atts );
+
+      
+      $remote_settings = Bizyhood_Utility::getRemoteSettings();
```
Comment (eugenebiz) [inline Bizyhood/Core.php]: The following block of code, to retrieve the configured API client, is repeated all over the place. The code should be refactored to use a common method to retrieve a client.

File: Bizyhood/Views/listings/promotions.php [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        $promotion['business_logo']['image']['url'] = Bizyhood_Utility::getImageBaseURL().'placeholder-logo.jpg';
+        $promotion['business_logo']['image_width'] = 307;
+        $promotion['business_logo']['image_height'] = 304;
+        
+        // check date
+        $start_date = date('Y-m-d', strtotime($promotion['start'])); // start date
+        $tomorrow_date = date('Y-m-d', strtotime('+ 1 days')); // tomorrow date
+        $end_date = date('Y-m-d', strtotime($promotion['end'])); // end date
+        
+        if (strtotime($start_date) < strtotime($tomorrow_date)) {
```
Comment (eugenebiz) [inline Bizyhood/Views/listings/promotions.php]: Why does this logic exist? I don't see any specification for it in the document for this widget.

File: Bizyhood/Views/listings/promotions.php [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        if ($i != 0 && ($i%2 == false)) {
+          ?>
+          </div><div class="bh_row bh_promotion-content promotions_page">	
+          <?php
+        }    
+        $i++;
+        
+        // set the default logo
+        $promotion['business_logo']['image']['url'] = Bizyhood_Utility::getImageBaseURL().'placeholder-logo.jpg';
+        $promotion['business_logo']['image_width'] = 307;
```
Comment (eugenebiz) [inline Bizyhood/Views/listings/promotions.php]: Any fixed values like width and height should be defined in a well-named constant, e.g. LOGO_WIDTH, so that we can change the values in a single place, should we need to.

File: Bizyhood/Views/listings/promotions.php [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+          } else {
+            $dates = '<span class="promotions_dates">Valid on '. date_i18n( get_option( 'date_format' ), strtotime($end_date)) .'</span>';
+          }
+        } else {
+          // display both start and ending day
+          $dates = '<span class="promotions_dates">Valid from '. date_i18n( get_option( 'date_format' ), strtotime($start_date)) .' to '. date_i18n( get_option( 'date_format' ), strtotime($end_date)) .'</span>';
+        }
+        
+        // trim the description if needed
+        if (str_word_count($promotion['details']) > 10) {
```
Comment (eugenebiz) [inline Bizyhood/Views/listings/promotions.php]: What is 10? Should be a well-named constant.

File: Bizyhood/Views/widgets/promotions.php [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+      // or exit with an error message
+      echo __('There are no promotions to display', 'bizyhood');
+      return;
+    }
+    
+    $view_business_page_id = get_page_by_path( "business-overview" )->ID;
+    
+    
+    // check date
+    $start_date = date('Y-m-d', strtotime($promotion['start'])); // start date
```
Comment (eugenebiz) [inline Bizyhood/Views/widgets/promotions.php]: All of this code was already written in the listings. Please refactor so that it uses a common template and PHP include as needed. We should minimize or eliminate duplication of code where possible.

File: Bizyhood/Core.php [hunk 6]
```diff
@@ -1033,10 +1070,129 @@ public function businesses_information($atts)
...
+      
+      return $return;
+      
+    }
+    
     /***** API Calls END *****/
     /***************************/
     
     
+    public function promotions_shortcode($attrs) {
```
Comment (eugenebiz) [inline Bizyhood/Core.php]: Core is getting unwieldy. Is there a sensible way to split it up by its respective functionality?

File: Bizyhood/Core.php [hunk 7]
```diff
@@ -930,9 +936,9 @@ public function businesses_information($atts)
       }
       
       // get verified
-      if (isset($a['verified'])) {
-          $verified = $a['verified'];
-      } elseif (get_query_var('verified')) {
+
+      $verified = $a['verified'];
```
Comment (eugenebiz) [inline Bizyhood/Core.php]: What is "a"? Please use variable names that give an idea of what they're representing at a glance.

File: Bizyhood/Core.php [hunk 8]
```diff
@@ -1033,16 +1018,95 @@ public function businesses_information($atts)
...
+     * API call for additional business info
+     * @param array $atts Attributes to be passesd on the API
+     * @param string $command data to get promotions (default) and events
+     * @return mixed results (array) or error (array) or empty
+     */
+    public function business_details_information($atts, $command = 'promotions')
+    {
+      
+      
+      $a = shortcode_atts( array(
```
Comment (eugenebiz) [inline Bizyhood/Core.php]: Ditto. Needs better variable name.

File: Bizyhood/Core.php [hunk 9]
```diff
@@ -1059,7 +1123,7 @@ public function businesses_shortcode($attrs)
        
         $businesses     = $q['businesses'];
         $keywords       = $q['keywords'];
-        $facets         = $q['cf'];
+        $facets         = $q['categories'];
```
Comment (eugenebiz) [inline Bizyhood/Core.php]: Why are $facets and $categories receiving the same value? Also, why do facets AND categories coexist? The facets represent the categories available in the search. How are they distinguished here?

File: Bizyhood/Core.php [hunk 10]
```diff
@@ -1078,6 +1142,44 @@ public function businesses_shortcode($attrs)
...
+    /**
+     * Sets and returns cached API results as a transient
+     * @param string $transient_name The name of the transient to get or set
+     * @param string $method_name The name of the function to call to get the API data
+     * @param array $attrs Attributes for the $method_name
+     * @param string $method_command The name of the command to be execute by the API
+     * @param boolean $random Either to return on random result or all of them
+     * @return array transient result(s) or false
+     */
+    public function try_transient($transient_name, $transient_value = 'response_json', $method_name, $attrs, $method_command = null, $random = false) {
```
Comment (eugenebiz) [inline Bizyhood/Core.php]: try_transient doesn't really tell me what this method is doing. Maybe get_cache_result would be a better start. Also, the transient_value parameter is misleading. It looks like that is a key name. You should probably change that to "transient_key".

File: Bizyhood/Views/listings/promotions.php (hunk 2 above)
Comment (eugenebiz) [inline Bizyhood/Views/listings/promotions.php]: Spyros, I don't see that the comment above has been addressed.

File: Bizyhood/Views/widgets/promotions.php (hunk 5 above)
Comment (eugenebiz) [inline Bizyhood/Views/widgets/promotions.php]: Spyros, the above comment has not been addressed as far as I can tell.

File: Bizyhood/oAuth.php [hunk 11]
```diff
@@ -0,0 +1,118 @@
+<?php
```
Comment (eugenebiz) [inline Bizyhood/oAuth.php]: Spyros, this is a great start to the refactor. I think we can get away with this for now. I would say that to make this class even more generic, you could override the constructor and offer the client settings as parameters. That way you abstract any notion of WP settings or the Bizyhood Utility from this class. This way the only dependency this class would have would be on the actual OAuth client.

Then, to be even more SOLID, you would abstract the OAuth client used entirely. Again, I don't think this is necessary in this pass. We can stick with what you have, these are just notes for your information.

File: Bizyhood/Core.php (hunk 7 above)
Comment (spyrosvl) [inline Bizyhood/Core.php]: It is common in WP to use $a for shortcode_atts values https://codex.wordpress.org/Shortcode_API . I can change the var name of course, something like $filtered_attributes?

File: Bizyhood/Core.php (hunk 8 above)
Comment (spyrosvl) [inline Bizyhood/Core.php]: I will come with a name that fits its purpose. I will ask you before using it though.

File: Bizyhood/Core.php (hunk 9 above)
Comment (spyrosvl) [inline Bizyhood/Core.php]: One of them is not needed any more. I will change the variables so it makes more sense.

File: Bizyhood/Core.php (hunk 10 above)
Comment (spyrosvl) [inline Bizyhood/Core.php]: I named it liek this, cause it will try to get the transient, but it may fail to do so. You are right about the $transient_value.

File: Bizyhood/Views/listings/promotions.php (hunk 2 above)
Comment (spyrosvl) [inline Bizyhood/Views/listings/promotions.php]: We need different text per case. Eg. we need "Valid on" or "Valid from...to..." etc. I had a discussion about it with Scott and consluded on this text. Should I change it ot put more comments on the code?

File: Bizyhood/Views/widgets/promotions.php (hunk 5 above)
Comment (spyrosvl) [inline Bizyhood/Views/widgets/promotions.php]: I think it is on my latest commit. I do not see it any more on my DEV.

File: Bizyhood/oAuth.php (hunk 11 above)
Comment (spyrosvl) [inline Bizyhood/oAuth.php]: I think I understand half of what you say, but it would be fo great help if we could do a share screen call about it :)

File: Bizyhood/Core.php (hunk 7 above)
Comment (eugenebiz) [inline Bizyhood/Core.php]: Yes, please do. I'd like to be able to look at a variable and have a basic understanding of what I'm looking at.

File: Bizyhood/Core.php (hunk 9 above)
Comment (eugenebiz) [inline Bizyhood/Core.php]: Okay, sounds good. Thanks.

File: Bizyhood/Views/widgets/promotions.php (hunk 5 above)
Comment (eugenebiz) [inline Bizyhood/Views/widgets/promotions.php]: This still looks like the same code, in part, as what's in: Bizyhood/Views/listings/promotions.php

File: Bizyhood/oAuth.php (hunk 11 above)
Comment (eugenebiz) [inline Bizyhood/oAuth.php]: We can take a look at this at a later time.

File: Bizyhood/Public/css/plugin.css [hunk 12]
```diff
@@ -35,12 +35,654 @@ a.facet_helper {
 .bh_row * {box-sizing:border-box;}
 .bh_row a {transition:.25s all linear;}
 
-.bh_col-md-1,.bh_col-md-2,.bh_col-md-3,.bh_col-md-4,.bh_col-md-5,.bh_col-md-6,.bh_col-md-7,.bh_col-md-8,.bh_col-md-9,.bh_col-md-10,.bh_col-md-11,.bh_col-md-12 {
+.bh_col-xs-1, .bh_col-sm-1, .bh_col-md-1, .bh_col-lg-1, .bh_col-xs-2, .bh_col-sm-2, .bh_col-md-2, .bh_col-lg-2, .bh_col-xs-3, .bh_col-sm-3, .bh_col-md-3, .bh_col-lg-3, .bh_col-xs-4, .bh_col-sm-4, .bh_col-md-4, .bh_col-lg-4, .bh_col-xs-5, .bh_col-sm-5, .bh_col-md-5, .bh_col-lg-5, .bh_col-xs-6, .bh_col-sm-6, .bh_col-md-6, .bh_col-lg-6, .bh_col-xs-7, .bh_col-sm-7, .bh_col-md-7, .bh_col-lg-7, .bh_col-xs-8, .bh_col-sm-8, .bh_col-md-8, .bh_col-lg-8, .bh_col-xs-9, .bh_col-sm-9, .bh_col-md-9, .bh_col-lg-9, .bh_col-xs-10, .bh_col-sm-10, .bh_col-md-10, .bh_col-lg-10, .bh_col-xs-11, .bh_col-sm-11, .bh_col-md-11, .bh_col-lg-11, .bh_col-xs-12, .bh_col-sm-12, .bh_col-md-12, .bh_col-lg-12
```
Comment (eugenebiz) [inline Bizyhood/Public/css/plugin.css]: Spyros, this looks like you're duplicating most of the grid layout classes from Bootstrap. Why is this necessary?

Comment (spyrosvl) [inline Bizyhood/Public/css/plugin.css]: Because the CSS as I found it when I started working on it, it only had half the classes. We need all the grid classes in order to make it look good on mobile too.

Comment (eugenebiz) [inline Bizyhood/Public/css/plugin.css]: So why not import the Bootstrap CSS instead of duplicating it in our own
CSS?

On Mon, Apr 25, 2016 at 9:56 PM, Spyros Vlachopoulos <
notifications@github.com> wrote:

> In Bizyhood/Public/css/plugin.css
> <https://github.com/Bizyhood/bizyhood-wp-plugin/pull/29#discussion_r61018811>
> :
>
> > @@ -35,12 +35,654 @@ a.facet_helper {
> >  .bh_row * {box-sizing:border-box;}
> >  .bh_row a {transition:.25s all linear;}
> >
> > -.bh_col-md-1,.bh_col-md-2,.bh_col-md-3,.bh_col-md-4,.bh_col-md-5,.bh_col-md-6,.bh_col-md-7,.bh_col-md-8,.bh_col-md-9,.bh_col-md-10,.bh_col-md-11,.bh_col-md-12 {
> > +.bh_col-xs-1, .bh_col-sm-1, .bh_col-md-1, .bh_col-lg-1, .bh_col-xs-2, .bh_col-sm-2, .bh_col-md-2, .bh_col-lg-2, .bh_col-xs-3, .bh_col-sm-3, .bh_col-md-3, .bh_col-lg-3, .bh_col-xs-4, .bh_col-sm-4, .bh_col-md-4, .bh_col-lg-4, .bh_col-xs-5, .bh_col-sm-5, .bh_col-md-5, .bh_col-lg-5, .bh_col-xs-6, .bh_col-sm-6, .bh_col-md-6, .bh_col-lg-6, .bh_col-xs-7, .bh_col-sm-7, .bh_col-md-7, .bh_col-lg-7, .bh_col-xs-8, .bh_col-sm-8, .bh_col-md-8, .bh_col-lg-8, .bh_col-xs-9, .bh_col-sm-9, .bh_col-md-9, .bh_col-lg-9, .bh_col-xs-10, .bh_col-sm-10, .bh_col-md-10, .bh_col-lg-10, .bh_col-xs-11, .bh_col-sm-11, .bh_col-md-11, .bh_col-lg-11, .bh_col-xs-12, .bh_col-sm-12, .bh_col-md-12, .bh_col-lg-12
>
> Because the CSS as I found it when I started working on it, it only had
> half the classes. We need all the grid classes in order to make it look
> good on mobile too.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/Bizyhood/bizyhood-wp-plugin/pull/29/files/b92b082250f1b08fdefdf225e402b4a1cedfc235#r61018811>
>



-- 
Eugene Fabrikant
Bizyhood Inc., Co-Founder & CTO
http://www.bizyhood.com | eugene@bizyhood.com
---
Build and be great

Comment (spyrosvl) [inline Bizyhood/Public/css/plugin.css]: 1. Because we do not need all of them.
2. We want the classes renamed, in case a theme has bootstrap and it is already tweaked.
3. Having 1 CSS file is better and faster than 2

Comment (eugenebiz) [inline Bizyhood/Public/css/plugin.css]: Spyros, I'll address your points:

1. Not needing all of them isn't a reason to break up the framework. Every
time we choose to use another piece of the bootstrap framework, we'd have
to bring in more sections of the framework, not just the actual pieces
used, but whatever is needed to support them. Doing so in piecemeal is not
only error prone from the manual copy perspective, but also from a
versioning perspective. We don't gain enough performance to warrant cherry
picking pieces of the framework and duplication of effort. Conversely, we
miss out on a stable, well-vetted, and widely supported framework.

2. If a theme utilizes Bootstrap, they shouldn't be modifying the framework
itself, they should be writing their own CSS classes to tweak as they see
fit, the same thing we should be doing in the plugin. In the horrible case
that they DO modify the framework itself, we would still be able to
compensate by writing our own classes, or worst case, recommend a different
theme, if and when possible. I'd rather encounter this problem and solve it
when it arises, rather than preempt it and miss out on the benefits of
having the bootstrap framework available to our plugin.

3. Given that contemporary bandwidth is on average more generous where are
target users are, and the fact that WP already loads a ridiculous amount of
dependencies since all plugins are loaded in every call, adding another CSS
file to the mix isn't going to impact the performance enough to warrant
such optimizations. I'd sooner recommend that the publishers remove
unnecessary plugins to save some cycles. Also, keep in mind that modern
browsers have increased the number of concurrent threads they have
available for loading resources and running tasks, in addition to the fact
that they cache resources once they've been downloaded once, unless
instructed otherwise by headers.

There are scenarios where one css file is not faster (from a user
perspective) than splitting into two. If you have smaller dependencies
loaded early, that allow the majority of the page to render mostly
correctly, that is a better experience than waiting for one large file that
has ALL dependencies, before the page can be rendered at all.

Needless to say, performance issues are not always obvious and definitely
should not be preempted, especially since in the WP universe, most
performance problems are caused either by crappy hosting, plugin overload,
the fact that it's difficult to control resources in the WP lifecycle,
(like our font loading bug), or just bad coding.

The benefits of building on top of mature, stable, and well supported
frameworks, far outweigh any performance issues we may incur because of
them. Let's have those performance issues first, then we'll see how to
correct them. Let's use CSS as it was intended first, before we start
hacking up the bootstrap framework.

On Tue, Apr 26, 2016 at 5:21 AM, Spyros Vlachopoulos <
notifications@github.com> wrote:

> In Bizyhood/Public/css/plugin.css
> <https://github.com/Bizyhood/bizyhood-wp-plugin/pull/29#discussion_r61053642>
> :
>
> > @@ -35,12 +35,654 @@ a.facet_helper {
> >  .bh_row * {box-sizing:border-box;}
> >  .bh_row a {transition:.25s all linear;}
> >
> > -.bh_col-md-1,.bh_col-md-2,.bh_col-md-3,.bh_col-md-4,.bh_col-md-5,.bh_col-md-6,.bh_col-md-7,.bh_col-md-8,.bh_col-md-9,.bh_col-md-10,.bh_col-md-11,.bh_col-md-12 {
> > +.bh_col-xs-1, .bh_col-sm-1, .bh_col-md-1, .bh_col-lg-1, .bh_col-xs-2, .bh_col-sm-2, .bh_col-md-2, .bh_col-lg-2, .bh_col-xs-3, .bh_col-sm-3, .bh_col-md-3, .bh_col-lg-3, .bh_col-xs-4, .bh_col-sm-4, .bh_col-md-4, .bh_col-lg-4, .bh_col-xs-5, .bh_col-sm-5, .bh_col-md-5, .bh_col-lg-5, .bh_col-xs-6, .bh_col-sm-6, .bh_col-md-6, .bh_col-lg-6, .bh_col-xs-7, .bh_col-sm-7, .bh_col-md-7, .bh_col-lg-7, .bh_col-xs-8, .bh_col-sm-8, .bh_col-md-8, .bh_col-lg-8, .bh_col-xs-9, .bh_col-sm-9, .bh_col-md-9, .bh_col-lg-9, .bh_col-xs-10, .bh_col-sm-10, .bh_col-md-10, .bh_col-lg-10, .bh_col-xs-11, .bh_col-sm-11, .bh_col-md-11, .bh_col-lg-11, .bh_col-xs-12, .bh_col-sm-12, .bh_col-md-12, .bh_col-lg-12
>
>
>    1. Because we do not need all of them.
>    2. We want the classes renamed, in case a theme has bootstrap and it
>    is already tweaked.
>    3. Having 1 CSS file is better and faster than 2
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/Bizyhood/bizyhood-wp-plugin/pull/29/files/b92b082250f1b08fdefdf225e402b4a1cedfc235#r61053642>
>



-- 
Eugene Fabrikant
Bizyhood Inc., Co-Founder & CTO
http://www.bizyhood.com | eugene@bizyhood.com
---
Build and be great

Comment (spyrosvl) [inline Bizyhood/Public/css/plugin.css]: Bootstrap is a generic framework. It will mess with everything that has to do with the site, eg inputs, forms, links. nav menus etc. There is also a possibility that a theme css class match with one of bootstrap. The results would be unexcpected ate least. My opinion has not changed. We should not load the full bootstrap framework or use classes of the same name. However, we can test it. So, for one last time; are you sure you want me to load bootstrap and convert all classes that we already use to the native bootstrap classes (no more bh_col-md-1, it will be col-md-1)?

Comment (spyrosvl) [inline Bizyhood/Public/css/plugin.css]: P.S. Bootstrap has a customizer: http://getbootstrap.com/customize/
So , everyone can tweak it before using it.

Comment (eugenebiz) [inline Bizyhood/Public/css/plugin.css]: Perfect. We can use that and update as needed. I'm pretty sure we will be
using more and more of Bootstrap as we expand the plugin. It's worked great
for us on our other web properties.

On Tue, Apr 26, 2016 at 12:28 PM, Spyros Vlachopoulos <
notifications@github.com> wrote:

> In Bizyhood/Public/css/plugin.css
> <https://github.com/Bizyhood/bizyhood-wp-plugin/pull/29#discussion_r61118248>
> :
>
> > @@ -35,12 +35,654 @@ a.facet_helper {
> >  .bh_row * {box-sizing:border-box;}
> >  .bh_row a {transition:.25s all linear;}
> >
> > -.bh_col-md-1,.bh_col-md-2,.bh_col-md-3,.bh_col-md-4,.bh_col-md-5,.bh_col-md-6,.bh_col-md-7,.bh_col-md-8,.bh_col-md-9,.bh_col-md-10,.bh_col-md-11,.bh_col-md-12 {
> > +.bh_col-xs-1, .bh_col-sm-1, .bh_col-md-1, .bh_col-lg-1, .bh_col-xs-2, .bh_col-sm-2, .bh_col-md-2, .bh_col-lg-2, .bh_col-xs-3, .bh_col-sm-3, .bh_col-md-3, .bh_col-lg-3, .bh_col-xs-4, .bh_col-sm-4, .bh_col-md-4, .bh_col-lg-4, .bh_col-xs-5, .bh_col-sm-5, .bh_col-md-5, .bh_col-lg-5, .bh_col-xs-6, .bh_col-sm-6, .bh_col-md-6, .bh_col-lg-6, .bh_col-xs-7, .bh_col-sm-7, .bh_col-md-7, .bh_col-lg-7, .bh_col-xs-8, .bh_col-sm-8, .bh_col-md-8, .bh_col-lg-8, .bh_col-xs-9, .bh_col-sm-9, .bh_col-md-9, .bh_col-lg-9, .bh_col-xs-10, .bh_col-sm-10, .bh_col-md-10, .bh_col-lg-10, .bh_col-xs-11, .bh_col-sm-11, .bh_col-md-11, .bh_col-lg-11, .bh_col-xs-12, .bh_col-sm-12, .bh_col-md-12, .bh_col-lg-12
>
> P.S. Bootstrap has a customizer: http://getbootstrap.com/customize/
> So , everyone can tweak it before using it.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/Bizyhood/bizyhood-wp-plugin/pull/29/files/b92b082250f1b08fdefdf225e402b4a1cedfc235#r61118248>
>



-- 
Eugene Fabrikant
Bizyhood Inc., Co-Founder & CTO
http://www.bizyhood.com | eugene@bizyhood.com
---
Build and be great
